### PR TITLE
[SPARK-19556][core] Do not encrypt block manager data in memory.

### PR DESCRIPTION
**Pull request by vanzin**
This change modifies the way block data is encrypted to make the more
common cases faster, while penalizing an edge case. As a side effect
of the change, all data that goes through the block manager is now
encrypted only when needed, including the previous path (broadcast
variables) where that did not happen.

The way the change works is by not encrypting data that is stored in
memory; so if a serialized block is in memory, it will only be encrypted
once it is evicted to disk.

The penalty comes when transferring that encrypted data from disk. If the
data ends up in memory again, it is as efficient as before; but if the
evicted block needs to be transferred directly to a remote executor, then
there's now a performance penalty, since the code now uses a custom
FileRegion implementation to decrypt the data before transferring.

This also means that block data transferred between executors now is
not encrypted (and thus relies on the network library encryption support
for secrecy). Shuffle blocks are still transferred in encrypted form,
since they're handled in a slightly different way by the code. This also
keeps compatibility with existing external shuffle services, which transfer
encrypted shuffle blocks, and avoids having to make the external service
aware of encryption at all.

The serialization and deserialization APIs in the SerializerManager now
do not do encryption automatically; callers need to explicitly wrap their
streams with an appropriate crypto stream before using those.

As a result of these changes, some of the workarounds added in SPARK-19520
are removed here.

Testing: a new trait ("EncryptionFunSuite") was added that provides an easy
way to run a test twice, with encryption on and off; broadcast, block manager
and caching tests were modified to use this new trait so that the existing
tests exercise both encrypted and non-encrypted paths. I also ran some
applications with encryption turned on to verify that they still work,
including streaming tests that failed without the fix for SPARK-19520.

### CODE REVIEW

**Comment by mridulm** [comment]
I have not looked at the implementation in detail, but can you comment on why the change w.r.t plain text block data to remote executor ? Isn't it not simpler to transmit block contents in encrypted format without decryption ?

Remote fetch of RDD blocks is not uncommon (for any task other than PROCESS_LOCAL); and I wanted to better understand why this is required.

**Comment by vanzin** [comment]
> Isn't it not simpler to transmit block contents in encrypted format without decryption?

First, keep in mind that there's no metadata that tells the receiver whether a block is encrypted or not. This means that methods like `BlockManager.get`, which can read block data from either local or remote sources, need to return data that is either always encrypted or always not encrypted for the same block ID.

This leaves two choices:

- encrypt the data in all stores (memory & disk); this is what the current code does, and it requires all code that uses the BlockManager to have to deal with encryption. This is what caused SPARK-19520, and I filed SPARK-19556 to cover yet another case of a code path that did not do the right thing when encryption is enabled.

- make all non-shuffle block data read from the BlockManager not encrypted. This means non-shuffle code calling the BlockManager does not have to care about encryption, since it will always read unencrypted data, and can always put unencrypted data in the BlockManager and it will be encrypted when needed (a.k.a. when writing to disk).

> Remote fetch of RDD blocks is not uncommon

That's fine. This change makes the data read from the BlockManager instance not encrypted. But when transmitting the data over to another executor, there's RPC-level encryption (`spark.authenticate.enableSaslEncryption` or `spark.network.crypto.enabled`), which means the data is still encrypted on the wire.

**Comment by mridulm** [comment]
> First, keep in mind that there's no metadata that tells the receiver whether a block is encrypted or not. This means that methods like BlockManager.get, which can read block data from either local or remote sources, need to return data that is either always encrypted or always not encrypted for the same block ID.

This can be solved by tagging the block data with a prefix byte - we do something similar for MapStatus (direct or broadcast).

**Comment by mridulm** [comment]
Just to be clear, I would prefer if we consistently did things - either encrypt all blocks while transferring (irrespective of sasl being enabled or not); or depend only on sasl for channel encryption.
But given this was what it currently is, I am not sure if it was by design or accident; and what the tradeoff's for ensuring consistency is.

(The workaround is, what I mentioned above, tagging)

**Comment by vanzin** [comment]
> This can be solved by tagging the block data with a prefix byte

Sure, it could be solved in different ways. I just happened to prefer the one in this patch, since I think it's less intrusive; if you look closely, the majority of changes are in a single class (`DiskStore`), and there's mostly minor adjustments in other places.

**Comment by vanzin** [comment]
> Just to be clear, I would prefer if we consistently did things - either encrypt all blocks while transferring (irrespective of sasl being enabled or not); or depend only on sasl for channel encryption.

Not really sure what you mean here. But transferring encrypted data without RPC encryption is not really secure, since the encryption key is transferred to executors using an RPC. There's even a warning message if RPC encryption is not on and you enable disk encryption.

Shuffle is a different beast - I explain why the shuffle blocks are transferred in encrypted form in the PR description.

**Comment by mridulm** [comment]
> Not really sure what you mean here. But transferring encrypted data without RPC encryption is not really secure, since the encryption key is transferred to executors using an RPC. There's even a warning message if RPC encryption is not on and you enable disk encryption.

Good point, I overlooked that.
So to summarize, after this change, RDD block's transferred will always be in plain text; with an implicit requirement that rpc encryption is strongly preferred to be enabled.
Is there any case where it is transfered in encrypted form in supported cases ? (cases being: broadcast, rdd block transfer, replication, anything else ?)
I wanted to ensure I understand what the final expected behavior/state would be, and how consistent we will become.

I agree about shuffle being special case'd; I was looking at only non-shuffle blocks.

**Comment by vanzin** [comment]
> Is there any case where it is transfered in encrypted form in supported cases ?

No, with these changes, only shuffle data is transferred in encrypted form.

**Comment by sameeragarwal** [comment]
cc @cloud-fan @ueshin

**Comment by cloud-fan** [comment]
> The penalty comes when transferring that encrypted data from disk. If the
data ends up in memory again, it is as efficient as before; but if the
evicted block needs to be transferred directly to a remote executor, then
there's now a performance penalty, since the code now uses a custom
FileRegion implementation to decrypt the data before transferring.

What's the actual difference? previously we transfer encrypted data?

**Comment by vanzin** [comment]
> What's the actual difference? previously we transfer encrypted data?

Yes. The previous version of the code would transfer the encrypted file over to the receiver, and the encrypted data for serialized blocks would also be stored in `MemoryStore` (and then decrypted on every use). That means the files could just be mmap'ed for transfer, which is faster than the `ReadableByteChannel` path even without encryption in the picture. (If you consider the previous code had to decrypt from the `MemoryStore` on every read, you can end up with better performance overall with this patch.)

But this caused all the other issues with making the `BlockManager` harder to use when encryption was on, so I think this is a better solution.

**Comment by cloud-fan** [comment]
makes sense. one more question, ideally, shall we also transfer shuffle blocks after decryption?

File: core/src/main/scala/org/apache/spark/security/CryptoStreamUtils.scala [hunk 1]
```diff
@@ -63,12 +83,40 @@ private[spark] object CryptoStreamUtils extends Logging {
...
       sparkConf: SparkConf,
       key: Array[Byte]): InputStream = {
-    val properties = toCryptoConf(sparkConf)
     val iv = new Array[Byte](IV_LENGTH_IN_BYTES)
-    is.read(iv, 0, iv.length)
-    val transformationStr = sparkConf.get(IO_CRYPTO_CIPHER_TRANSFORMATION)
-    new CryptoInputStream(transformationStr, properties, is,
-      new SecretKeySpec(key, "AES"), new IvParameterSpec(iv))
+    var read = 0
+    while (read < iv.length) {
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/security/CryptoStreamUtils.scala]: what does this while loop do?

File: core/src/main/scala/org/apache/spark/security/CryptoStreamUtils.scala [hunk 2]
```diff
@@ -102,4 +150,34 @@ private[spark] object CryptoStreamUtils extends Logging {
     }
     iv
   }
+
+  /**
+   * This class is a workaround for CRYPTO-125, that forces all bytes to be written to the
+   * underlying channel. Since the callers of this API are using blocking I/O, there are no
+   * concerns with regards to CPU usage here.
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/security/CryptoStreamUtils.scala]: is it a separated bug fix?

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 3]
```diff
@@ -1235,7 +1251,7 @@ private[spark] class BlockManager(
           peer.port,
           peer.executorId,
           blockId,
-          new NettyManagedBuffer(data.toNetty),
+          new BlockManagerManagedBuffer(blockInfoManager, blockId, data.toManagedBuffer()),
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: why this change?

File: core/src/main/scala/org/apache/spark/storage/DiskStore.scala [hunk 4]
```diff
@@ -73,55 +86,219 @@ private[spark] class DiskStore(conf: SparkConf, diskManager: DiskBlockManager) e
...
+    override def size(): Long = blockSize
+
+    override def nioByteBuffer(): ByteBuffer = {
+      // This is used by the block transfer service to replicate blocks. The upload code reads
+      // all bytes into memory to send the block to the remote executor, so it's ok to do this
+      // as long as the block fits in a Java array.
+      assert(blockSize <= Int.MaxValue, "Block is too large to be wrapped in a byte buffer.")
+      val is = toInputStream()
+      try {
+        ByteBuffer.wrap(ByteStreams.toByteArray(is))
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/storage/DiskStore.scala]: will we read all data out here?

**Comment by vanzin** [comment]
>  shall we also transfer shuffle blocks after decryption?

No. That's explained in the PR description.

File: core/src/main/scala/org/apache/spark/storage/DiskStore.scala (hunk 4 above)
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/storage/DiskStore.scala]: There's a comment explaining it a few lines above...

File: core/src/main/scala/org/apache/spark/security/CryptoStreamUtils.scala (hunk 1 above)
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/security/CryptoStreamUtils.scala]: It avoids issues with short reads. It's unlikely to happen but I always write read code like this to be safe.

File: core/src/main/scala/org/apache/spark/security/CryptoStreamUtils.scala (hunk 2 above)
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/security/CryptoStreamUtils.scala]: No. As the comment states, it's a workaround for a bug in the commons-crypto library, which would affect the code being added.

File: core/src/main/scala/org/apache/spark/security/CryptoStreamUtils.scala (hunk 1 above)
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/security/CryptoStreamUtils.scala]: Yeah, you can just use `ByteStreams.readFully(is, iv)`.

Comment (vanzin) [inline core/src/main/scala/org/apache/spark/security/CryptoStreamUtils.scala]: Ah, missed that one. +1 for shorter code.

File: core/src/main/scala/org/apache/spark/storage/DiskStore.scala [hunk 5]
```diff
@@ -73,55 +86,219 @@ private[spark] class DiskStore(conf: SparkConf, diskManager: DiskBlockManager) e
...
+    try {
+      var remaining = blockSize
+      val chunks = new ListBuffer[ByteBuffer]()
+      while (remaining > 0) {
+        val chunkSize = math.min(remaining, Int.MaxValue)
+        val chunk = allocator(chunkSize.toInt)
+        remaining -= chunkSize
+
+        while (chunk.remaining() > 0) {
+          source.read(chunk)
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/storage/DiskStore.scala]: as mentioned above, needs EOF error handling.

File: core/src/main/scala/org/apache/spark/storage/DiskStore.scala [hunk 6]
```diff
@@ -73,55 +86,219 @@ private[spark] class DiskStore(conf: SparkConf, diskManager: DiskBlockManager) e
...
+      } else {
+        lastWrite = 0
+      }
+    }
+
+    _transferred += written
+    written
+  }
+
+  override def deallocate(): Unit = source.close()
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/storage/DiskStore.scala]: release buffer as well.

**Review by mridulm — COMMENTED** [review]
I did an initial pass and added some comments/queries. Overall, as I mentioned earlier, I like the fact that we have a more consistent approach to transfering data.
Thanks for the work @vanzin !

File: core/src/main/scala/org/apache/spark/security/CryptoStreamUtils.scala [hunk 7]
```diff
@@ -102,4 +150,34 @@ private[spark] object CryptoStreamUtils extends Logging {
     }
     iv
   }
+
+  /**
+   * This class is a workaround for CRYPTO-125, that forces all bytes to be written to the
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/security/CryptoStreamUtils.scala]: There's a pretty nasty workaround for it in the network library... (the non-blocking workaround is a lot worse than this.)

File: core/src/main/scala/org/apache/spark/storage/DiskBlockManager.scala [hunk 8]
```diff
@@ -34,6 +34,8 @@ import org.apache.spark.util.{ShutdownHookManager, Utils}
  */
 private[spark] class DiskBlockManager(conf: SparkConf, deleteFilesOnStop: Boolean) extends Logging {
 
+  private val METADATA_FILE_SUFFIX = ".meta"
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/storage/DiskBlockManager.scala]: Hmm, good point... there's currently no metadata kept in the `DiskStore` class, but then this shouldn't be a lot of data.

File: core/src/main/scala/org/apache/spark/storage/DiskStore.scala [hunk 9]
```diff
@@ -17,48 +17,61 @@
...
     var threwException: Boolean = true
     try {
-      writeFunc(fileOutputStream)
+      writeFunc(out)
+      Files.write(out.getCount().toString(), diskManager.getMetadataFile(blockId), UTF_8)
       threwException = false
     } finally {
       try {
-        Closeables.close(fileOutputStream, threwException)
+        Closeables.close(out, threwException)
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/storage/DiskStore.scala]: This was the previous behavior, but well, doesn't hurt to fix it.

File: core/src/main/scala/org/apache/spark/storage/DiskStore.scala [hunk 10]
```diff
@@ -73,55 +86,219 @@ private[spark] class DiskStore(conf: SparkConf, diskManager: DiskBlockManager) e
...
   def putBytes(blockId: BlockId, bytes: ChunkedByteBuffer): Unit = {
-    put(blockId) { fileOutputStream =>
-      val channel = fileOutputStream.getChannel
-      Utils.tryWithSafeFinally {
-        bytes.writeFully(channel)
-      } {
-        channel.close()
-      }
+    put(blockId) { channel =>
+      bytes.writeFully(channel)
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/storage/DiskStore.scala]: The channel is owned by the code in the `put` method, which does that.

File: core/src/main/scala/org/apache/spark/storage/DiskStore.scala [hunk 11]
```diff
@@ -73,55 +86,219 @@ private[spark] class DiskStore(conf: SparkConf, diskManager: DiskBlockManager) e
...
+
+  private def openForWrite(file: File): WritableByteChannel = {
+    val out = new FileOutputStream(file).getChannel()
+    try {
+      securityManager.getIOEncryptionKey().map { key =>
+        CryptoStreamUtils.createWritableChannel(out, conf, key)
+      }.getOrElse(out)
+    } catch {
+      case e: Exception =>
+        out.close()
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/storage/DiskStore.scala]: There might be exceptions specific for the commons-crypto library being thrown.

File: core/src/main/scala/org/apache/spark/storage/DiskStore.scala (hunk 6 above)
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/storage/DiskStore.scala]: `StorageUtils.dispose` specifically checks for mapped buffers, which is not the case here. It could be changed, but in this case I wonder if it's necessary or if waiting for GC is good enough.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 12]
```diff
@@ -56,6 +57,43 @@ private[spark] class BlockResult(
...
+    val buffer: ChunkedByteBuffer,
+    autoDispose: Boolean = true) extends BlockData {
+
+  override def toInputStream(): InputStream = buffer.toInputStream(dispose = autoDispose)
+
+  override def toManagedBuffer(): ManagedBuffer = new NettyManagedBuffer(buffer.toNetty)
+
+  override def toByteBuffer(allocator: Int => ByteBuffer): ChunkedByteBuffer = {
+    buffer.copy(allocator)
+  }
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: So I had traced through that stuff 2 or 3 times, and now I did it again and I think I finally understood all that's going on. Basically, the old code was really bad at explicitly disposing of the buffers, meaning a bunch of paths (like the ones that used managed buffers) didn't bother to do it and just left the work to the GC.

I changed the code a bit to make the dispose more explicit and added comments in a few key places.

**Comment by vanzin** [comment]
retest this please

File: core/src/main/scala/org/apache/spark/broadcast/TorrentBroadcast.scala [hunk 13]
```diff
@@ -219,18 +219,22 @@ private[spark] class TorrentBroadcast[T: ClassTag](obj: T, id: Long)
...
+              blocks.map(_.toInputStream()), SparkEnv.get.serializer, compressionCodec)
+            // Store the merged copy in BlockManager so other tasks on this executor don't
+            // need to re-fetch it.
+            val storageLevel = StorageLevel.MEMORY_AND_DISK
+            if (!blockManager.putSingle(broadcastId, obj, storageLevel, tellMaster = false)) {
+              throw new SparkException(s"Failed to store $broadcastId in BlockManager")
+            }
+            obj
+          } finally {
+            blocks.foreach(_.dispose())
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/broadcast/TorrentBroadcast.scala]: ah good catch! we should dispose the blocks here

File: core/src/main/scala/org/apache/spark/security/CryptoStreamUtils.scala [hunk 14]
```diff
@@ -48,12 +51,30 @@ private[spark] object CryptoStreamUtils extends Logging {
...
+   * Wrap a `WritableByteChannel` for encryption.
+   */
+  def createWritableChannel(
+      channel: WritableByteChannel,
+      sparkConf: SparkConf,
+      key: Array[Byte]): WritableByteChannel = {
+    val params = new CryptoParams(key, sparkConf)
+    val iv = createInitializationVector(params.conf)
+    val buf = ByteBuffer.wrap(iv)
+    while (buf.hasRemaining()) {
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/security/CryptoStreamUtils.scala]: is there any possibility this may be an infinite loop?

File: core/src/main/scala/org/apache/spark/security/CryptoStreamUtils.scala [hunk 15]
```diff
@@ -63,12 +84,27 @@ private[spark] object CryptoStreamUtils extends Logging {
...
+  /**
+   * Wrap a `ReadableByteChannel` for decryption.
+   */
+  def createReadableChannel(
+      channel: ReadableByteChannel,
+      sparkConf: SparkConf,
+      key: Array[Byte]): ReadableByteChannel = {
+    val iv = new Array[Byte](IV_LENGTH_IN_BYTES)
+    val buf = ByteBuffer.wrap(iv)
+    JavaUtils.readFully(channel, buf)
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/security/CryptoStreamUtils.scala]: why not use `ByteStreams.readFully`? the `buf` is not used else where

File: core/src/main/scala/org/apache/spark/security/CryptoStreamUtils.scala (hunk 14 above)
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/security/CryptoStreamUtils.scala]: actually this logic is same with `CryptoHelperChannel`. Shall we create `CryptoHelperChannel` first and simply call `helper.write(buf)` here?

File: core/src/main/scala/org/apache/spark/serializer/SerializerManager.scala [hunk 16]
```diff
@@ -167,30 +167,26 @@ private[spark] class SerializerManager(
     val byteStream = new BufferedOutputStream(outputStream)
     val autoPick = !blockId.isInstanceOf[StreamBlockId]
     val ser = getSerializer(implicitly[ClassTag[T]], autoPick).newInstance()
-    ser.serializeStream(wrapStream(blockId, byteStream)).writeAll(values).close()
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/serializer/SerializerManager.scala]: the `wrapStream` and `wrapForEncryption` methods can be removed from this class

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 17]
```diff
@@ -56,6 +57,44 @@ private[spark] class BlockResult(
...
+ */
+private[spark] trait BlockData {
+
+  def toInputStream(): InputStream
+
+  def toNetty(): Object
+
+  def toChunkedByteBuffer(allocator: Int => ByteBuffer): ChunkedByteBuffer
+
+  def toByteBuffer(): ByteBuffer
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: it will be great to add some document for these 4 methods about when they will be called.

File: core/src/main/scala/org/apache/spark/storage/BlockManagerManagedBuffer.scala [hunk 18]
```diff
@@ -17,11 +17,15 @@
...
+import java.io.InputStream
+import java.nio.ByteBuffer
+import java.util.concurrent.atomic.AtomicInteger
+
+import org.apache.spark.network.buffer.ManagedBuffer
 import org.apache.spark.util.io.ChunkedByteBuffer
 
 /**
- * This [[ManagedBuffer]] wraps a [[ChunkedByteBuffer]] retrieved from the [[BlockManager]]
+ * This [[ManagedBuffer]] wraps a ManagedBuffer retrieved from the [[BlockManager]]
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerManagedBuffer.scala]: `wraps a [[BlockData]]`

File: core/src/main/scala/org/apache/spark/storage/BlockManagerManagedBuffer.scala [hunk 19]
```diff
@@ -31,17 +35,31 @@ import org.apache.spark.util.io.ChunkedByteBuffer
 private[storage] class BlockManagerManagedBuffer(
     blockInfoManager: BlockInfoManager,
     blockId: BlockId,
-    chunkedBuffer: ChunkedByteBuffer) extends NettyManagedBuffer(chunkedBuffer.toNetty) {
+    data: BlockData,
+    dispose: Boolean) extends ManagedBuffer {
+
+  private val refCount = new AtomicInteger(1)
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerManagedBuffer.scala]: maybe we should mention it in the class doc that the `BlockData` will be disposed automatically via reference count.

File: core/src/main/scala/org/apache/spark/storage/BlockManagerManagedBuffer.scala [hunk 20]
```diff
@@ -31,17 +35,31 @@ import org.apache.spark.util.io.ChunkedByteBuffer
 private[storage] class BlockManagerManagedBuffer(
     blockInfoManager: BlockInfoManager,
     blockId: BlockId,
-    chunkedBuffer: ChunkedByteBuffer) extends NettyManagedBuffer(chunkedBuffer.toNetty) {
+    data: BlockData,
+    dispose: Boolean) extends ManagedBuffer {
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerManagedBuffer.scala]: `needDispose` may be a better name

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 21]
```diff
@@ -1065,7 +1084,7 @@ private[spark] class BlockManager(
           try {
             replicate(blockId, bytesToReplicate, level, remoteClassTag)
           } finally {
-            bytesToReplicate.unmap()
+            bytesToReplicate.dispose()
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: why change `unmap` to `dispose`?

File: core/src/main/scala/org/apache/spark/storage/DiskStore.scala [hunk 22]
```diff
@@ -17,48 +17,67 @@
...
-      writeFunc(fileOutputStream)
+      writeFunc(out)
+      blockSizes.put(blockId.name, out.getCount)
       threwException = false
     } finally {
       try {
-        Closeables.close(fileOutputStream, threwException)
+        out.close()
+      } catch {
+        case ioe: IOException =>
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/storage/DiskStore.scala]: why this? `threwException` starts with `true`

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 23]
```diff
@@ -56,6 +57,44 @@ private[spark] class BlockResult(
...
 
 /**
+ * Abstracts away how blocks are stored and provides different ways to read the underlying block
+ * data. Callers should call [[dispose()]] when they're done with the block.
+ */
+private[spark] trait BlockData {
+
+  def toInputStream(): InputStream
+
+  def toNetty(): Object
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: why the return type is `Object`?

File: core/src/main/scala/org/apache/spark/security/CryptoStreamUtils.scala (hunk 14 above)
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/security/CryptoStreamUtils.scala]: No, there's no infinite loop here, because a failure would cause an exception. Yeah, using the helper should work too.

File: core/src/main/scala/org/apache/spark/security/CryptoStreamUtils.scala (hunk 15 above)
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/security/CryptoStreamUtils.scala]: There's no `ByteStreams.readFully` for `ReadableByteChannel` that I'm aware of.

File: core/src/main/scala/org/apache/spark/serializer/SerializerManager.scala (hunk 16 above)
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/serializer/SerializerManager.scala]: They're still used in a bunch of places.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala (hunk 21 above)
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: Because there's no `BlockData.unmap()`.

File: core/src/main/scala/org/apache/spark/storage/BlockManagerManagedBuffer.scala (hunk 20 above)
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerManagedBuffer.scala]: Hmm, I prefer `dispose`, because it's not about needing to dispose the buffer, but wanting to dispose the buffer.

File: core/src/main/scala/org/apache/spark/storage/DiskStore.scala (hunk 22 above)
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/storage/DiskStore.scala]: The code needs to catch any exception thrown by `out.close()` and also remove the block in that case. That wasn't done before.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala (hunk 23 above)
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: See `ManagedBuffer.convertToNetty()`.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala (hunk 17 above)
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: I added scaladoc for `toNetty()`, but the others seem self-explanatory to me.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala (hunk 21 above)
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: I'm afraid this may counteract the effort we made in https://github.com/apache/spark/pull/16499

Ideally `unmap` and `dispose` do different things

cc @mallman

Comment (vanzin) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: `BlockData.dispose` calls `ChunkedByteBuffer.unmap`.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 24]
```diff
@@ -56,6 +57,49 @@ private[spark] class BlockResult(
...
+
+  override def toChunkedByteBuffer(allocator: Int => ByteBuffer): ChunkedByteBuffer = {
+    buffer.copy(allocator)
+  }
+
+  override def toByteBuffer(): ByteBuffer = buffer.toByteBuffer
+
+  override def size: Long = buffer.size
+
+  override def dispose(): Unit = buffer.unmap()
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: can we define the semantic of the `BlockData.dispose` clearly? It's quite confusing here that the `dispose` method call `buffer.unmap` while `ChunkedByteBuffer` also has a `dispose` method.

Comment (vanzin) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: I think `BlockData.dispose()` is pretty well defined. "Release any resources held by the object." What's confusing is that there's both `dispose()` and `unmap()` in `ChunkedByteBuffer`, when there used to be only `dispose()`. It's confusing to have two different methods for releasing resources, and that confusion is not being caused by this patch.

`BlockData` is not just a wrapper around `ChunkedByteBuffer`; if it were there wouldn't be a need for it. Which is why calling the method `unmmap()` wouldn't make any sense here, since that's very specific to memory-mapped byte buffers.

Comment (vanzin) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: BTW I'm really starting to think the fix in #16499, while technically correct, is more confusing that it should be. The problem is not that the code was disposing of off-heap buffers; the problem is that buffers read from the memory store should not be disposed of, while buffers read from the disk store should.

So it's not really a matter of dispose vs. unmap, but a matter of where the buffer come from. (Which is kinda what I had in this patch with the `autoDispose` parameter to `ByteBufferBlockData`. Perhaps I should revive that and get rid of `StorageUtils.unmap`, which is just confusing.)

**Comment by vanzin** [comment]
I removed `StorageUtils.unmap()` in my last commit (see commit message for details). That makes the confusion go away.

The replication tests fail from time to time but they seem to be flaky without this patch. See;
https://spark-tests.appspot.com/test-details?suite_name=org.apache.spark.storage.BlockManagerProactiveReplicationSuite&test_name=proactive+block+replication+-+5+replicas+-+4+block+manager+deletions

File: core/src/main/scala/org/apache/spark/storage/BlockManagerManagedBuffer.scala [hunk 25]
```diff
@@ -17,31 +17,52 @@
...
+import org.apache.spark.network.buffer.ManagedBuffer
 import org.apache.spark.util.io.ChunkedByteBuffer
 
 /**
- * This [[ManagedBuffer]] wraps a [[ChunkedByteBuffer]] retrieved from the [[BlockManager]]
+ * This [[ManagedBuffer]] wraps a [[BlockData]] instance retrieved from the [[BlockManager]]
  * so that the corresponding block's read lock can be released once this buffer's references
  * are released.
  *
+ * If `dispose` is set to try, the [[BlockData]]will be disposed when the buffer's reference
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerManagedBuffer.scala]: `is set to try` -> `is set to true`

**Comment by cloud-fan** [comment]
LGTM, cc @mallman to check the `unmap` part

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by mallman** [comment]
> LGTM, cc @mallman to check the unmap part

LGTM, too. Sorry for the late reply... I've been away the past two weeks.
